### ISSUE TITLE
Two-way bindings gives 'Cannot save value from target back to source' in debug output

### ISSUE THREAD

**Issue by isaacabraham** [elmish/Elmish.WPF#10]
On a simple two-way binding from a textbox to a string on a model, I see the following output when running: -

```
System.Windows.Data Error: 8 : Cannot save value from target back to source. BindingExpression:Path=Name; DataItem='ViewModelBase`2' (HashCode=64479624); target element is 'TextBox' (Name=''); target property is 'Text' (type 'String') InvalidOperationException:'System.InvalidOperationException: Property path is not valid. 'System.Dynamic.DynamicObject+MetaDynamic' does not have a public property named 'Items'.
   at CallSite.Target(Closure , CallSite , Object , Object )
   at MS.Internal.DynamicPropertyAccessorImpl.SetValue(Object component, Object value)
   at MS.Internal.Data.PropertyPathWorker.SetValue(Object item, Object value)
   at MS.Internal.Data.ClrBindingWorker.UpdateValue(Object value)
   at System.Windows.Data.BindingExpression.UpdateSource(Object value)'
```

This doesn't crash the app, but repeated errors like these have an effort on WPF perf when they mount up.

**Comment by 2sComplement** [comment]
Are you using `Binding.twoWay` in your viewBindings?

**Comment by isaacabraham** [comment]
Yes, this is a two-way binding.

**Comment by 2sComplement** [comment]
Do you have `Mode=TwoWay` in your XAML binding? If so and you're still seeing this, I'd need to see a code snippet to reproduce.

**Comment by isaacabraham** [comment]
Here's the repo: https://github.com/CompositionalIT/ElmishWpfSample

(Note: `Mode=TwoWay` is the default on XAML bindings in WPF, so it's not set explicitly.)

**Comment by 2sComplement** [comment]
I see it now and know why it's happening. It shouldn't affect any functionality but you're right about effects on performance. I will look into a way to avoid this.

**Comment by BentTranberg** [comment]
I see this in the Elmish.Samples.Counter demo, or to be exact, in my own demo where the Elmish.Samples.Counter demo source was copied, to get a start. It happens every time the slider is moved.

Just mentioning it in case there's need for a repro.

Error message:

```
System.Windows.Data Error: 8 : Cannot save value from target back to source. BindingExpression:Path=StepSize; DataItem='ViewModelBase`2' (HashCode=32361769); target element is 'Slider' (Name=''); target property is 'Value' (type 'Double') InvalidOperationException:'System.InvalidOperationException: Property path is not valid. 'System.Dynamic.DynamicObject+MetaDynamic' does not have a public property named 'Items'.
   at CallSite.Target(Closure , CallSite , Object , Object )
   at MS.Internal.DynamicPropertyAccessorImpl.SetValue(Object component, Object value)
   at MS.Internal.Data.PropertyPathWorker.SetValue(Object item, Object value)
   at MS.Internal.Data.ClrBindingWorker.UpdateValue(Object value)
   at System.Windows.Data.BindingExpression.UpdateSource(Object value)'
```

**Comment by BentTranberg** [comment]
@isaacabraham, I have one case now where Mode=TwoWay has to be set explicitly, or else it won't work. But I have also seen that in most cases it appears to be the default.

**Comment by cmeeren** [comment]
I also experience this bug. Will it ever be fixed? :)

**Comment by cmeeren** [comment]
@2sComplement in #15 you said:

> The ViewModel is trying to prevent an immediate get after the view sets the binding source

Why are you preventing the get?

**Comment by cmeeren** [comment]
After having looked into this a bit, it does indeed seem like view performs an immediate `get` after the value has been set. This `get` will in most (all?) cases return the existing value, not the new value, because the loop hasn't had a chance to complete yet.

This leaves me with three questions:

1. What is the actual performance impact of this error?
2. Is the performance impact relevant outside of Debug mode?
3. Is there anything to be done about this? I haven't come up with a good solution on the ViewModel side, and I'm not that experienced with WPF so I don't know if it can be fixed from that end (e.g. if there is some setting that disables the immediate `get` or simply turns of the error).

**Comment by cmeeren** [comment]
Oh, another question, perhaps the most important - can @2sComplement or anyone else answer me this: Why is the `get` prevented in the first place? I have tried returning `true`, and it doesn't seem to have an impact on the functionality.

**Comment by 2sComplement** [comment]
If I recall correctly, allowing it resulted in some undesirable view behaviour. For example, typing in an input box would be annoying, assuming you wanted characters to appear in the correct order.  With each key press the view would update the text box with the old value before updating the model.

**Comment by cmeeren** [comment]
Thanks for the quick reply! Yes, I observe the behavior you're describing; I initially only tested with a slider. With a TextBox however, you essentially type backwards.

That leaves my first three questions.

**Comment by 2sComplement** [comment]
Apologies for my absence in general lately, been very tied up on other projects. I couldn't prove any real impacts of this error.  Nor do I know a way around it without somehow overriding the behaviour of WPF through perhaps some custom binding type, which is probably the route I would have taken to investigate a fix for this. Not convinced it's worth it though.

**Comment by JDiLenarda** [comment]
I noticed this afternoon that on aBinding.oneWay, the getter is called 3 times per update (on a silly example involving a button that increased an integer and a textbox displaying it). I was wondering if that was due to your implementation or if it was already the case before, but I didn’t check. Is it the case your talking about ?

Naive suggestion for the get called just after the set : would a timed out lock or mutex on TryGetMember and TrySetmember be helpful ?

I’m far from my office PC right now, I can’t watch exactly what the problem is.

**Comment by cmeeren** [comment]
@JDiLenarda it's not the same if it's a one-way binding, no.

Please take a look at the SingleCounter example in the rewrite branch, TryGetMember is only called once there. If you find out why it's called three times in your example, let us know, and let us know if it's a bug or if we should make something clear in the documentation. If you can't figure it out, I can also take a repro solution, but I'm busy this week with other projects too, so I can't promise I can look at it right away.

As for locking: I'm very wary of doing that due to general multithreaded complexity and other performance concerns following from that solution, and also because I'm not sure the reported error actually matters at all outside development/debugging scenarios.

**Comment by JDiLenarda** [comment]
Now I think about it, it’s probably because the Textbox is on Mode=TwoWay by default. I’ll check tomorrow.

**Comment by cmeeren** [comment]
Thanks for the clarification, @JDiLenarda (didn't see your edit before now).

To everyone and no-one in particular, allow me to quote my three unanswered questions lest they get lost in the thread. Perhaps @ReedCopsey knows the answer to some of these questions?

> 1. What is the actual performance impact of the OP error?
> 2. Is the performance impact relevant outside of Debug mode?
> 3. Is there anything to be done about this? I haven't come up with a good solution on the ViewModel side, and I'm not that experienced with WPF so I don't know if it can be fixed from that end (e.g. if there is some setting that disables the immediate get or simply turns of the error message).

**Comment by cmeeren** [comment]
This should be fixed in 2.0.0-beta-10
